From 183fc10543a956143d1b61fd876ba65ea0bcfa76 Mon Sep 17 00:00:00 2001
From: Frank van der Linden <fllinden@amazon.com>
Date: Sat, 10 Nov 2018 00:18:32 +0000
Subject: [PATCH] xen: Only restore the ACPI SCI interrupt in
 xen_restore_pirqs.

Restoring all PIRQs, which is the right thing to do, was causing problems
on larger instances. This is a horrible workaround until this issue is fully
understood.

[Srinidhi Rao <srinidhi.rao@broadcom.com> ported this patch to v6.6.x]

Signed-off-by: Frank van der Linden <fllinden@amazon.com>
Reviewed-by: Alakesh Haloi <alakeshh@amazon.com>
Reviewed-by: Anchal Agarwal <anchalag@amazon.com>
Reviewed-by: Qian Lu <luqia@amazon.com>
---
 drivers/xen/events/events_base.c | 42 +++++++++++++++++++++++++++++++-
 1 file changed, 41 insertions(+), 1 deletion(-)

diff --git a/drivers/xen/events/events_base.c b/drivers/xen/events/events_base.c
index 913563dbc5d1..3f1b588536a8 100644
--- a/drivers/xen/events/events_base.c
+++ b/drivers/xen/events/events_base.c
@@ -68,6 +68,10 @@
 #include <xen/xenbus.h>
 #include <asm/hw_irq.h>

+#ifdef CONFIG_ACPI
+#include <linux/acpi.h>
+#endif
+
 #include "events_internal.h"

 #undef MODULE_PARAM_PREFIX
@@ -2123,9 +2127,45 @@ void xen_shutdown_pirqs(void)
 	}
 }

+/*
+ * For now, only restore the ACPI SCI pirq.
+ */
 void xen_restore_pirqs(void)
 {
-	restore_pirqs();
+#ifdef CONFIG_ACPI
+	int pirq, rc, irq, gsi;
+	struct physdev_map_pirq map_irq;
+	struct irq_info *info;
+
+	list_for_each_entry(info, &xen_irq_list_head, list) {
+		if (info->type != IRQT_PIRQ)
+			continue;
+
+		pirq = info->u.pirq.pirq;
+		gsi = info->u.pirq.gsi;
+		irq = info->irq;
+
+		if (gsi != acpi_gbl_FADT.sci_interrupt)
+			continue;
+
+		map_irq.domid = DOMID_SELF;
+		map_irq.type = MAP_PIRQ_TYPE_GSI;
+		map_irq.index = gsi;
+		map_irq.pirq = pirq;
+
+		rc = HYPERVISOR_physdev_op(PHYSDEVOP_map_pirq, &map_irq);
+		if (rc) {
+			pr_warn("xen: ACPI SCI interrupt map failed, rc=%d\n",
+				rc);
+			xen_free_irq(irq);
+			continue;
+		}
+
+		printk(KERN_DEBUG "xen: restored ACPI SCI interrupt\n");
+
+		__startup_pirq(irq);
+	}
+#endif
 }

 static struct irq_chip xen_dynamic_chip __read_mostly = {
--
2.35.6

